reference/src/layout/integers-floatingpoint.md

```diff
@@ -37,6 +37,25 @@ Note: on all platforms that Rust's currently supports, the size and alignment of
 
 For full ABI compatibility details, see [Gankro’s post](https://gankro.github.io/blah/rust-layouts-and-abis/#the-layoutsabis-of-builtins).
 
+## Fixed-width integer types
+
+Rust's fixed-width integer types `{i,u}{8,16,32,64}` have the same layout as the
+C fixed-width integer types from the `<stdint.h>` header
+`{u,}int{8,16,32,64}_t`. That is:
+
+* these types have no padding bits,
+* their size exactly matches their bit-width,
+* negative values of signed integer types are represented using `2`'s complement.
+
+Therefore these integer types are safe to use directly in C FFI where the
+fixed-width integer types are expected.
+
+## Fixed-width floating point types
+
+Rust's `f32` and `f64` types have the same layout as C's `float` and `double`
+types, respectively. Therefore these floating-point types are safe to use
+directly in C FFI where the appropriate C types are expected.
```

Comment:
I don't think we can state that in such generality, at least not unilaterally. I believe C allows implementing `float` and `double` as the same format, so that `double` *also* corresponds to `f32` (this is potentially useful if you target a platform with f32 hardware but no f64 support). Annex F would not apply to such an implementation, but as long as the implementation doesn't define `__STDC_IEC_559__` as 1 that's conforming.

Reply:
@rkruppe AFAICT, if `__STDC_IEC_559__` is not defined, there are no guarantees whatsoever about what `float` and `double` could be (except that `double` is at least as large as `float`).

Maybe I should rewrite this to say that `f32` and `f64` only matches `float` and `double` if `<math.h>` defines `__STDC_IEC_559__`, and that otherwise they don't necessarily match. 

Do all platforms that Rust currently support define `__STDC_IEC_559__`? We could add some info about this as a note.

Reply:
I am a bit hesitant to recommand tying outselves to `__STDC_IEC_559__`, because although there is a long line of official sentiments that Rust's floating point types are always IEEE 754, it is definitely possible for a C implementation to provide Rust-compatible floating point types without implementing the rest of Annex F (e.g., floating point environment access). Indeed Clang is currently in that situation (though in practice the `math.h` you get defines `__STDC_IEC_559__` and so it's technically not conforming).